server/schedule/merge_checker.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/pingcap/pd/server/namespace"
+	log "github.com/sirupsen/logrus"
+)
+
+// MergeChecker ensures region to merge with adjacent region when size is small
+type MergeChecker struct {
+	cluster    Cluster
+	classifier namespace.Classifier
+}
+
+// NewMergeChecker creates a merge checker.
+func NewMergeChecker(cluster Cluster, classifier namespace.Classifier) *MergeChecker {
+	return &MergeChecker{
+		cluster:    cluster,
+		classifier: classifier,
+	}
+}
+
+// Check verifies a region's replicas, creating an Operator if need.
+func (m *MergeChecker) Check(region *core.RegionInfo) (*Operator, *Operator) {
+	// region size is not small enough
+	if region.ApproximateSize >= int64(m.cluster.GetMaxMergeRegionSize()) {
+		return nil, nil
+	}
+
+	// skip hot region
+	if m.cluster.IsRegionHot(region.GetId()) {
+		return nil, nil
+	}
+
+	var target *core.RegionInfo
+	prev, next := m.cluster.GetAdjacentRegions(region)
+
+	peerCount := len(region.Region.GetPeers())
+	var direction metapb.MergeDirection
+	// if is not hot region and under same namesapce
+	if prev != nil && !m.cluster.IsRegionHot(prev.GetId()) && m.classifier.AllowMerge(region, prev) {
+		// peer count should equal
+		if peerCount == len(prev.Region.GetPeers()) {
+			target = prev
+			direction = metapb.MergeDirection_Backward
+		}
+	}
+	if next != nil && !m.cluster.IsRegionHot(next.GetId()) && m.classifier.AllowMerge(region, next) {
+		// if both region is not hot, prefer the one with smaller size
+		if target == nil || target.ApproximateSize > next.ApproximateSize {
+			// peer count should equal
+			if peerCount == len(next.Region.GetPeers()) {
+				target = next
+				direction = metapb.MergeDirection_Forward
+			}
+		}
+	}
+
+	if target == nil {
+		return nil, nil
+	}
+
+	steps, err := m.matchPeers(region, target)
+	if err != nil {
+		return nil, nil
+	}
+
+	log.Debugf("try to merge region {%v} into region {%v}", region, target)
+	op1, op2 := CreateMergeRegionOperator("merge-region", region, target, direction, OpMerge, steps)
+
+	return op1, op2
+}
+
+func (m *MergeChecker) matchPeers(source *core.RegionInfo, target *core.RegionInfo) ([]OperatorStep, error) {
+	storeIDs := make(map[uint64]struct{})
+	var steps []OperatorStep
+
+	sourcePeers := source.Region.GetPeers()
+	targetPeers := target.Region.GetPeers()
+
+	for _, peer := range targetPeers {
+		storeIDs[peer.GetStoreId()] = struct{}{}
+	}
+
+	// Add missing peers.
+	for id := range storeIDs {
+		if source.GetStorePeer(id) != nil {
+			continue
+		}
+		peer, err := m.cluster.AllocPeer(id)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		steps = append(steps, AddPeer{ToStore: id, PeerID: peer.Id})
+	}
+
+	// Check whether to transfer leader or not
+	intersection := m.getIntersectionStores(sourcePeers, targetPeers)
+	leaderID := source.Leader.GetStoreId()
+	isFound := false
+	for _, storeID := range intersection {
+		if storeID == leaderID {
+			isFound = true
+			break
+		}
+	}
+	if isFound == false {
```

Comment:
!isFound

Reply:
I'm confused. Why doesn't it need to transfer leader when leader is one of the intersect stores?

Reply:
leader is not need to be in same store